riscv/processor.cc

```diff
@@ -217,144 +217,120 @@ void processor_t::parse_priv_string(const char* str)
 
 void processor_t::parse_isa_string(const char* str)
 {
-  std::string lowercase = strtolower(str), tmp;
-
-  char error_msg[256];
-  const char* p = lowercase.c_str();
+  isa_string = strtolower(str);
   const char* all_subsets = "imafdqchp"
 #ifdef __SIZEOF_INT128__
     "v"
 #endif
     "";
 
-  max_xlen = 64;
   max_isa = reg_t(2) << 62;
+  if (isa_string.compare(0, 4, "rv32") == 0)
+    max_xlen = 32, max_isa = reg_t(1) << 30;
+  else if (isa_string.compare(0, 4, "rv64") == 0)
+    max_xlen = 64;
+  else
+    bad_isa_string(str, "Spike supports either RV32I or RV64I");
+  if (isa_string[3] == 'g')
+    isa_string = isa_string.substr(0, 4) + "imafd" + isa_string.substr(5);
+  if (isa_string[4] != 'i')
+    bad_isa_string(str, "'I' extension is required");
 
-  if (strncmp(p, "rv32", 4) == 0)
-    max_xlen = 32, max_isa = reg_t(1) << 30, p += 4;
-  else if (strncmp(p, "rv64", 4) == 0)
-    p += 4;
-  else if (strncmp(p, "rv", 2) == 0)
-    p += 2;
-
-  if (!*p) {
-    p = "imafdc";
-  } else if (*p == 'g') { // treat "G" as "IMAFD"
-    tmp = std::string("imafd") + (p+1);
-    p = &tmp[0];
+  auto p = isa_string.begin();
+  for (p += 4; islower(*p) && !strchr("zshx", *p); ++p) {
+    while (*all_subsets && (*p != *all_subsets))
+      ++all_subsets;
+    if (!*all_subsets)
+      bad_isa_string(str, "Wrong order");
+    switch (*p) {
+      case 'q': max_isa |= 1L << ('d' - 'a');
```

Comment:
This needs to be `(1L << ('d' - 'a')) | (1L << ('f' - 'a'))`, because Q implies D then D implies Q.  I'll fix it as part of testing it out, though.

Reply:
Nevermind, missed that this is a fallthrough...  good ol' C.